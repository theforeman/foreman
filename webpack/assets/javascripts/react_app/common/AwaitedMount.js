import PropTypes from 'prop-types';
import { useState, useEffect, useCallback } from 'react';
import store from '../redux';
import componentRegistry from '../components/componentRegistry';

// Mounts a component after all plugins have been imported to make sure that all plugins are available to the component
export const AwaitedMount = ({ component, data, flattenData }) => {
  const [mounted, setMounted] = useState(false);
  const [mountedComponent, setMountedComponent] = useState(null);
  const mountComponent = useCallback(async () => {
    if (componentRegistry.registry[component]) {
      setMounted(true);
      setMountedComponent(
        componentRegistry.markup(component, {
          data,
          store,
          flattenData,
        })
      );
    } else {
      // eslint-disable-next-line no-console
      console.debug(
        `Component not found: ${component}. The script for the component might not have been loaded yet.`
      );
    }
  }, [component, data, flattenData]);
  useEffect(() => {
    document.addEventListener('loadPlugin', mountComponent);
    return () => {
      window.removeEventListener('loadPlugin', mountComponent);
    };
  }, [mountComponent]);
  useEffect(() => {
    mountComponent();
  }, [mountComponent]);
  return mounted ? mountedComponent : null;
};

AwaitedMount.propTypes = {
  component: PropTypes.string.isRequired,
  data: PropTypes.object,
  flattenData: PropTypes.bool,
};
AwaitedMount.defaultProps = {
  data: {},
  flattenData: false,
};
