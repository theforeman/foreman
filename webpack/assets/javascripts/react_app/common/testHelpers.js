import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

export default {
  mockStorage: () => {
    const storage = {};

    return {
      setItem: (key, value) => {
        storage[key] = value || '';
      },
      getItem: key => (key in storage ? storage[key] : null),
      removeItem: key => {
        delete storage[key];
      },
      get length() {
        return Object.keys(storage).length;
      },
      key: i => {
        const keys = Object.keys(storage);

        return keys[i] || null;
      },
    };
  },
};

// a helper method for invoking a class method (for unit tests)
// obj = a class
// func = a tested function
// objThis = an object's this
// arg = function args

export const classFunctionUnitTest = (obj, func, objThis, args) =>
  obj.prototype[func].apply(objThis, args);

/**
 * Shallow render a component multipile times with fixtures
 * @param  {ReactComponent} Component Component to shallow-render
 * @param  {Object}         fixtures  key=fixture description, value=props to apply
 * @return {Object}                   key=fixture description, value=shallow-rendered component
 */
export const shallowRenderComponentWithFixtures = (Component, fixtures) =>
  Object.entries(fixtures).map(([description, props]) => ({
    description,
    component: shallow(<Component {...props} />),
  }));

/**
 * Test a component with fixtures and snapshots
 * @param  {ReactComponent} Component Component to test
 * @param  {Object}         fixtures  key=fixture description, value=props to apply
 */
export const testComponentSnapshotsWithFixtures = (Component, fixtures) =>
  shallowRenderComponentWithFixtures(Component, fixtures).forEach(
    ({ description, component }) =>
      it(description, () => expect(toJson(component)).toMatchSnapshot())
  );

/**
 * run an action (sync or async) and except the results to much snapshot
 * @param  {Function}  runAction  Action runner function
 * @return {Promise}
 */
export const testActionSnapshot = async runAction => {
  const actionResults = runAction();

  // if it's an async action
  if (typeof actionResults === 'function') {
    const dispatch = jest.fn();
    await actionResults(dispatch);

    expect(dispatch.mock.calls).toMatchSnapshot();
  } else {
    expect(actionResults).toMatchSnapshot();
  }
};

/**
 * Test actions with fixtures and snapshots
 * @param  {Object} fixtures key=fixture description, value=action runner function
 */
export const testActionSnapshotWithFixtures = fixtures =>
  Object.entries(fixtures).forEach(([description, runAction]) =>
    it(description, () => testActionSnapshot(runAction))
  );

/**
 * Test a reducer with fixtures and snapshots
 * @param  {Function} reducer  reducer to test
 * @param  {Object}   fixtures key=fixture description, value=props to apply
 */
export const testReducerSnapshotWithFixtures = (reducer, fixtures) => {
  const reduce = ({ state, action = {} } = {}) => reducer(state, action);
  Object.entries(fixtures).forEach(([description, action]) =>
    it(description, () => expect(reduce(action)).toMatchSnapshot())
  );
};

/**
 * Test selectors with fixtures and snapshots
 * @param  {Object} fixtures  key=fixture description,
 *                            value=selector runner function
 */
export const testSelectorsSnapshotWithFixtures = fixtures =>
  Object.entries(fixtures).forEach(([description, selectorRunner]) =>
    it(description, () => expect(selectorRunner()).toMatchSnapshot())
  );
