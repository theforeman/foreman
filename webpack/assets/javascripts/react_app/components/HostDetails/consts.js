export const DEFAULT_TAB = 'Overview';
export const HOST_DETAILS_KEY = 'HOST_DETAILS';
export const HOST_DETAILS_API_OPTIONS = { key: HOST_DETAILS_KEY };
export const TABS_SLOT_ID = 'host-details-page-tabs';
