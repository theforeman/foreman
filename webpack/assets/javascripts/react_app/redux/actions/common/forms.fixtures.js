export const requestData = {
  values: { a: 1 },
  url: '/api/resource',
  item: 'Resource',
};

export const requestDataMsg = Object.assign({}, requestData, {
  message: 'Customized success!',
});
