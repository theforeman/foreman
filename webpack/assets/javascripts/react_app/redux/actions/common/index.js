import API from '../../../API';

export const ajaxRequestAction = ({
  dispatch,
  requestAction,
  successAction,
  failedAction,
  url,
  item,
}) => {
  dispatch({ type: requestAction, payload: item });
  return API.get(url)
    .then(({ data }) =>
      dispatch({ type: successAction, payload: { ...item, ...data } })
    )
    .catch(error => dispatch({ type: failedAction, payload: { error, item } }));
};
