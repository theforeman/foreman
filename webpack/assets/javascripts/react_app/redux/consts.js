export const TOASTS_ADD = 'TOASTS_ADD';
export const TOASTS_CLEAR = 'TOASTS_CLEAR';
export const TOASTS_DELETE = 'TOASTS_DELETE';
export const VMWARE_CLUSTER_CHANGE = 'VMWARE_CLUSTER_CHANGE';
export const STORAGE_VMWARE_INIT = 'STORAGE_VMWARE_INIT';
export const STORAGE_VMWARE_ADD_CONTROLLER = 'STORAGE_VMWARE_ADD_CONTROLLER';
export const STORAGE_VMWARE_ADD_DISK = 'STORAGE_VMWARE_ADD_DISK';
export const STORAGE_VMWARE_REMOVE_CONTROLLER =
  'STORAGE_VMWARE_REMOVE_CONTROLLER';
export const STORAGE_VMWARE_UPDATE_CONTROLLER =
  'STORAGE_VMWARE_UPDATE_CONTROLLER';
export const STORAGE_VMWARE_REMOVE_DISK = 'STORAGE_VMWARE_REMOVE_DISK';
export const STORAGE_VMWARE_UPDATE_DISK = 'STORAGE_VMWARE_UPDATE_DISK';
export const NOTIFICATIONS = 'NOTIFICATIONS';
export const STORAGE_VMWARE_DATASTORES = 'STORAGE_VMWARE_DATASTORES';
export const STORAGE_VMWARE_DATASTORES_REQUEST =
  'STORAGE_VMWARE_DATASTORES_REQUEST';
export const STORAGE_VMWARE_DATASTORES_SUCCESS =
  'STORAGE_VMWARE_DATASTORES_SUCCESS';
export const STORAGE_VMWARE_DATASTORES_FAILURE =
  'STORAGE_VMWARE_DATASTORES_FAILURE';
export const STORAGE_VMWARE_STORAGEPODS = 'STORAGE_VMWARE_STORAGEPODS';
export const STORAGE_VMWARE_STORAGEPODS_REQUEST =
  'STORAGE_VMWARE_STORAGEPODS_REQUEST';
export const STORAGE_VMWARE_STORAGEPODS_SUCCESS =
  'STORAGE_VMWARE_STORAGEPODS_SUCCESS';
export const STORAGE_VMWARE_STORAGEPODS_FAILURE =
  'STORAGE_VMWARE_STORAGEPODS_FAILURE';
export const NOTIFICATIONS_TOGGLE_DRAWER = 'NOTIFICATIONS_TOGGLE_DRAWER';
export const NOTIFICATIONS_SET_EXPANDED_GROUP =
  'NOTIFICATIONS_SET_EXPANDED_GROUP';
export const NOTIFICATIONS_MARK_AS_READ = 'NOTIFICATIONS_MARK_AS_READ';
export const NOTIFICATIONS_MARK_GROUP_AS_READ =
  'NOTIFICATIONS_MARK_GROUP_AS_READ';
export const NOTIFICATIONS_MARK_AS_CLEAR = 'NOTIFICATIONS_MARK_AS_CLEAR';
export const NOTIFICATIONS_MARK_GROUP_AS_CLEARED =
  'NOTIFICATIONS_MARK_GROUP_AS_CLEARED';
export const NOTIFICATIONS_LINK_CLICKED = 'NOTIFICATIONS_LINK_CLICKED';
export const PASSWORD_STRENGTH_PASSWORD_CHANGED =
  'PASSWORD_STRENGTH_PASSWORD_CHANGED';
export const PASSWORD_STRENGTH_PASSWORD_MATCHED =
  'PASSWORD_STRENGTH_PASSWORD_MATCHED';
