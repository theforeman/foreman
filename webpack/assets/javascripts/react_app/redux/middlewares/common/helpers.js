/**
 * Executes a callback when the document is visible.
 * Used to decrease load when tab is hidden, for example when intervals are running.
 * @param { Function } callback
 */
export const whenDocumentIsVisible = callback => {
  const { hidden, msHidden, webkitHidden } = document;
  let isHidden = true;
  const isNotUndefined = n => typeof n !== 'undefined';

  if (isNotUndefined(hidden)) {
    // Opera 12.10 and Firefox 18 and later support
    isHidden = hidden;
  } else if (isNotUndefined(msHidden)) {
    isHidden = msHidden;
  } else if (isNotUndefined(webkitHidden)) {
    isHidden = webkitHidden;
  }

  if (!isHidden) {
    callback();
  }
};
