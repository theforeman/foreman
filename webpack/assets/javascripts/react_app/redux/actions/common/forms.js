import { APIActions } from '../../API';
import { sprintf, translate as __ } from '../../../../react_app/common/I18n';

export class SubmissionError {
  constructor(errors) {
    this.errors = errors;
  }
}

const fieldErrors = ({ error }) => {
  const { errors, severity } = error;

  if (errors.base) {
    errors._error = {};
    errors._error.errorMsgs = errors.base;
    errors._error.severity = severity;
    delete errors.base;
  }

  return new SubmissionError(errors);
};

export const onError = error => {
  if (error.response.status === 422) {
    // Handle invalid form data
    throw fieldErrors(error.response.data);
  }
  throw new SubmissionError({
    _error: {
      errorMsgs: [
        `${__('Error submitting data:')} ${error.response.status} ${__(
          error.response.statusText
        )}`,
      ],
    },
  });
};

const verifyProps = (item, values) => {
  if (!item) {
    throw new Error('item must be defined, e.g. Bookmark');
  }
  if (!values) {
    throw new Error('values must be defined');
  }
};

export const submitForm = ({
  item,
  url,
  values: params,
  message,
  method = 'post',
  headers,
  apiActionTypes: actionTypes,
}) => {
  verifyProps(item, params);
  return async dispatch => {
    const uniqueAPIKey = `${item.toUpperCase()}_FORM_SUBMITTED`;

    const handleError = error => onError(error);

    const handleSuccess = ({ data }) => {
      dispatch({
        type: uniqueAPIKey,
        payload: { item, data },
      });
    };

    const successToast = response =>
      message || sprintf('%s was successfully created.', __(item));

    const errorToast = error =>
      sprintf(
        'Oh no! Something went wrong while submitting the form, the server returned the following error: %s',
        error
      );

    dispatch(
      APIActions[method]({
        key: uniqueAPIKey,
        url,
        headers,
        params,
        actionTypes,
        handleError,
        handleSuccess,
        successToast,
        errorToast,
      })
    );
  };
};
