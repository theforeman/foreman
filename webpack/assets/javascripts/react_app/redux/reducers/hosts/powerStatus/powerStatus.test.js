import * as types from '../../../consts';

import {
  initialState,
  request,
  stateBeforeResponse,
  response,
  error,
} from './powerStatus.fixtures';

import reducer from './index';
import { testReducerSnapshotWithFixtures } from '../../../../common/testHelpers';

describe('powerStatus reducer', () => {
  const fixtures = {
    'should return the initial state': {
      state: undefined,
      action: {},
    },
    'should handle HOST_POWER_STATUS_REQUEST': {
      state: initialState,
      action: {
        type: types.HOST_POWER_STATUS_REQUEST,
        payload: request,
      },
    },
    'should handle HOST_POWER_STATUS_SUCCESS': {
      state: stateBeforeResponse,
      action: {
        type: types.HOST_POWER_STATUS_SUCCESS,
        payload: response,
      },
    },
    'should handle HOST_POWER_STATUS_FAILURE': {
      state: stateBeforeResponse,
      action: {
        type: types.HOST_POWER_STATUS_FAILURE,
        payload: { error, item: { id: request.id } },
      },
    },
  };
  testReducerSnapshotWithFixtures(reducer, fixtures);
});
