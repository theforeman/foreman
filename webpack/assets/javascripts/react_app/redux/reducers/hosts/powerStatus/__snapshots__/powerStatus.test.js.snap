// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`powerStatus reducer should handle HOST_POWER_STATUS_FAILURE 1`] = `
Object {
  "2": Object {
    "error": "some error happened",
  },
}
`;

exports[`powerStatus reducer should handle HOST_POWER_STATUS_REQUEST 1`] = `
Object {
  "2": Object {
    "id": "2",
    "url": "test",
  },
}
`;

exports[`powerStatus reducer should handle HOST_POWER_STATUS_SUCCESS 1`] = `
Object {
  "2": Object {
    "data": "data",
    "id": "2",
  },
}
`;

exports[`powerStatus reducer should return the initial state 1`] = `Object {}`;
